### DISCUSSION TITLE
Setting to Turn Off "Fast Refresh" and to stop Next.js from automatically reloading the page against my will

### DISCUSSION THREAD

**Discussion by meglio** [vercel/next.js#25712] [Ideas]
### Describe the feature you'd like to request

The ability to turn Fast Refresh OFF and stop Next.js from refreshing my page automatically against my will.

### Describe the solution you'd like

A configuration setting that would disable Fast Refresh and let me refresh the page in browser manually.

### Describe alternatives you've considered

Dear team, some 3rd party libraries are not compatible with Fast Refresh. I cannot a) ask every library developer to fix their libraries to work with Next.js Fast Refresh; and b) wait for them to complete the updates.

Why can't we just switch Fast Refresh OFF with a configuration option and move on with using custom 3rd party packages with Next.js?

Please reopen this issue - it is an issue as it's simply not a viable option to get all libraries respect requirements imposed by the Fast Refresh feature of Next.js

For instance, I'd like to turn it off entirely for the whole project as I am using a 3rd party in-house library that is not compatible with Fast Refresh and will not be made so in the next year at least, if at all.

----

The second argument against Fast Refresh is that it is not sophisticated enough to preserve the state when I need it preserved, and it keeps refreshing the page against my will. I just want to stop it and keep hitting F5 or Command-R to refresh the page in the browser manually.

The experience of the page being updated against my will when I don't need it is frustrating.

**Comment by meglio** [comment]
Hey team, why did a feature request become a discussion? It's that simple, make a way to let Next.js stop refreshing the page - what can be wrong about having a way to control this simple behavior? What kind of discussion do you expect? Do you want to find people who'll say that's not an issue? Of course there are people who don't need this setting.

**Comment by simon360** [comment]
Have you looked into [customising Next's webpack config](https://nextjs.org/docs/api-reference/next.config.js/custom-webpack-config)? I expect you could find the fast refresh plugin/config in there and disable it.

**Reply by meglio** [reply]
@simon360 could you help with that? it's undocumented, so at least a minimal hint / example is needed in documentation.

It is weird that I have to opt out of special functionality instead of opting in for it.

I can understand Next.js refreshing the page when the source code changes. But just refreshing it automatically every N seconds is making me freak out sometimes. Actually, quite often:

- when debugging
- when analyzing updates to the database made with every page refresh
- when inspecting my HTML in browser, after all!

There are adult people who can hit Command+R / F5 when they need to. Please, just make it possible to turn this automatic refresh every N seconds OFF, otherwise it stays in the way of working efficiently.

**Reply by chipit24** [reply]
I tried disabling it with the following in my `next.config.js` file:
```js
module.exports = {
  /* ... */
  webpack(config) {
    config.plugins = config.plugins.filter(plugin => {
      return plugin.constructor.name !== 'ReactFreshWebpackPlugin';
    });

    return config;
  },
};
```

However, this broke my app with the following error:
```
Uncaught ReferenceError: $RefreshSig$ is not defined
```

I didn't spend much time with it so I'm sure there's something obvious I'm missing (editing the babel config, etc.). But +1 for getting a documented way to disable fast refresh / HMR.

**Reply by benkingcode** [reply]
@meglio there is no "refresh every N seconds" as far as I'm aware

**Comment by kevinseabourne** [comment]
I was looking to disable fast refresh also. I have my prettier format on save, and If I don't type fast enough. It formats due to fast refresh on every change and my code is all over the place. I ended up just disabling prettier (format on save) and have it set to manual with the keymap the same as formatting on save before fast refresh (cmd + s) 
seems way better, my code is running on every change and I can format whenever I want.

**Comment by ivanmaeder** [comment]
I've come to this because of what I thought may have been a [bug](https://github.com/vercel/next.js/issues/29184) but is apparently the expected behaviour. Also I think I have a slightly different perspective to @meglio's original post.

Right now I'm working on a Stripe integration and I'm having some issues with automatic reloading:

- A page that uses `getServerSideProps()` and has side effects: I really don't want this page to run out of the blue (some setup is required and if it runs before the setup is finished I have to start again)
- A webhook which—I think—sometimes refreshes just as it's being invoked (I can't control when it gets called but I guess even if I could, I can't always predict when automatic reloading might run)

The general feeling of not being in control is very frustrating and I'd rather sacrifice automatic reloading completely, at least in certain scenarios.

Thanks 👍

**Reply by siscodev93** [reply]
I was working with an api and got kicked after refreshing so many times. I have a tendency to ctrl+s every couple of key strokes - this hit the api at least 500 times since every reload counts for 3 hits? Regardless I'm glad it was a free access point if it was AWS I woulda racked up a bill :/

**Comment by Pickra** [comment]
Hey @meglio, I'm having the same problem. Still trying to debug to see if there's anything I can do w/my code or at least identify the cause. You mention...
> ...some 3rd party libraries are not compatible with Fast Refresh.

Are you referring to the Fast Refresh [limitations](https://nextjs.org/docs/basic-features/fast-refresh#limitations)?
Any details will be helpful.

Thank you.

**Reply by meglio** [reply]
Hi @Pickra , in my particular case, I'm using an internal corporate library that servers multiple websites with multi-million traffic. I have no option to stop using that library.

**Reply by Pickra** [reply]
Thank you for responding @meglio . I specifically want to know, what you mean by
> some 3rd party libraries are not compatible with Fast Refresh.

Do you mean this? - https://nextjs.org/docs/basic-features/fast-refresh#limitations

**Reply by meglio** [reply]
Yeah, those libraries simply cannot satisfy all the requirements listed in that Fast Refresh article.

**Reply by Pickra** [reply]
Thanks again @meglio. I know people who love nextjs but it ain't working for my app so I ripped it out.

**Comment by meglio** [comment]
New `[Fast Refresh] rebuilding` entries in developer console are displayed time to time without me touching a single file of the codebase. It is very confusing:

1) I have no idea why it rebuilds
2) I cannot turn this auto-rebuild off
3) I simply cannot stop it doing something without me asking it to do it

It is very dissatisfying and makes for bad user experience.

Please, please, give us an option to turn fast refresh off entirely.

It would also be very useful if it added a comment to the log entries in console about WHY exactly it rebuilt - was it because some file got changed (if so, which one?), or was it on a timer (if so, then how do we turn that timer off), etc.

These unpredictable page updates are killing developer experience entirely in some scenarios. Please can you do something about it. I can't understand how you dedicate so much resources on collaboration and super features like in V12, but still can't sort basic stuff like this out. Please could you pay attention to this issue - it is a serious deal-breaker.

**Comment by keaton-codes-tech** [comment]
I can't believe there is no option to turn this off. Really? I'm very close to just going back to Node and Vue. Developer experience with Next has been frustrating.

**Reply by mehzee92** [reply]
> I can't believe there is no option to turn this off. Really? I'm very close to just going back to Node and Vue. Developer experience with Next has been frustrating.

same feeling.

**Reply by zallesov** [reply]
Same here. I struggle with next more than developing

**Reply by Schachte** [reply]
Same. Turning into corporate spaghetti

**Comment by MedDimessi** [comment]
I have the same issue, I was testing a login page with form and submit button, and to figure out any bugs or misbehavior, the page automatically reload and the console log clears up and I can't read what happened and what logs I've put to understand what's going on!!
It's not a solution as I am here looking for efficient one, but I put `alert("the object or the log I am trying to display")` and it stops the reload, and then I can be able to read the console history.
But it is still frustrating as the trying to control the lifecycle of the app after submitting is hard to debug because of the random reloading and clearing up the console.

**Reply by benkingcode** [reply]
Just enable `Preserve console log` in your DevTools?

**Reply by MedDimessi** [reply]
What about the app state that get cleared?? like I said, it's not a solution, just what I thought at the time to help me debug.

**Reply by ScyDev** [reply]
@dbbk That doesn't address the problem at all.

**Reply by OFD16** [reply]
The year is 2025 and I have the same problem. There was no such problem in registration but almost the same code but there is in login and I am trying to determine whether there is a problem with the API or the frontend by putting an alert for hours like a joke. I tried a lot of things to prevent automatic refresh but it didn't work. I have developed projects using next.js before but I did not encounter such a problem because I made my own components. But when I used the simple html form structure it started to rebuild itself it is driving me crazy

**Comment by murilozilli** [comment]
Changing the environment variable NODE_ENV='production' for me seems to fix it

**Reply by ScyDev** [reply]
But there's a reason for dev mode. Developers don't want to work in production mode.

**Reply by murilozilli** [reply]
I know, just informing of an alternative to at least get things going as apparently that was a blocker for some people

**Comment by ScyDev** [comment]
Same problem here. This doesn't look good at all for Next.js.

Same as @meglio, I just leave my Chrome and Firefox alone with the page open, and not changing any code files at all. Every N seconds, the page will do a full refresh with, saying on the console:
```
[Fast Refresh] rebuilding
next-dev.js?c2df:52 Refreshing page data due to server-side change
```

**Reply by MedDimessi** [reply]
Idk if we have the same issue exactly, but I solved mine by something that has nothing to do with nextjs fast refresh, I had a submit form, when submitting, the event that is linked to the submit button reloads the page. so, when I set `e.preventDefault()` before my logic and it worked like a charm.

**Comment by ScyDev** [comment]
Is it possible that this could perhaps be caused by the IDE or a Git client, that is changing some file in the app dir, causing Next Js to do a fast refresh because it listens to some file system signal?

Does anyone have a clue how we could get more info an WHICH file change exactly is triggering Next?

**Comment by ScyDev** [comment]
One small workaround I found that sometimes helps: pause JS execution in the dev tools of your browser. This way, at least you can look at the console log or DOM elements without Next refreshing on you for no good reason.

**Comment by pnede3** [comment]
Please make it possible to turn off periodic refreshing.

The app I'm working on for a customer isn't built as correctly as you expect, and the registration wizard is built on a single page. Whenever I'm in page 5 and the auto-refresh kicks in, it automatically goes back to slide 1.

**Comment by ycnmhd** [comment]
I had a similar issue where the UI would refresh when a call to an api route `/api/some-route` is made
In the CLI console I get this:
```
wait  - compiling /api/some-route (server only)...
event - compiled successfully in 1169 ms (1787 modules)
```
After a lot of frustration, I found that the issue happens on a page that uses `getServerSideProps`, from which I’m calling a function that is also used by `/api/some-route`.
By removing the import of that shared function, and calling `/api/some-route` directly within `getServerSideProps` using `fetch`, ([which is not recommended in the documentation](https://nextjs.org/docs/basic-features/data-fetching/get-server-side-props#getserversideprops-or-api-routes)) the issue was solved.

**Comment by ScyDev** [comment]
Found the solution for my specific case by accident: There was a component that tried to act like a page (probably a dev did this out of ignorance). 

It was passed `lazyProps` as one of its `props`, used them with `useLazyProps()` to track a state and used `createLazyProps()` to try and populate its initialProps.

All in all, this doesn't make sense to me and I have removed it all, using just a normal react state inside the component. 
`[Fast Refresh]` still seems to happen and is printed on the console, but it doesn't cause a full reload of the page anymore, which was the annoying part.

**Comment by anees-synavos** [comment]
Unnamed components seems to be a problem

**Comment by Joxter** [comment]
I have a page with complex animations with a long of `.addEventListener` and sometimes hot reload freeze and crash the whole browser tab

<img width="267" alt="image" src="https://user-images.githubusercontent.com/3627930/184815735-f184c259-0f4c-4c29-a0d4-bcc5387553ad.png">

every event have a proper `.removeEventListener`

**Comment by m-lautenbach** [comment]
I would just like to turn hot reloading off because we get a chunk loading error when navigating between pages and I want to check if the hot reloading is causing that.

Not sure why NextJS is so weird about these things, SSR can't be easily turned off either.
It's like these things are not built in a modular way in their code base and that's why they resist providing config options to turn them off...

**Comment by parrisvarney** [comment]
It appears hot reloading causes a page refresh even while you have an active breakpoint in Firefox debugger, +1 to a way to turn it off.

**Comment by aptarmy** [comment]
There are some workarounds.

You can prevent page from full refresh on client side, this will ask you to confirm everytime before page is going to be refreshed.

**preventFastRefreshDev.js**
```
import { useEffect } from 'react';

const confirmPageRefresh = e => {
  e.preventDefault();
  const shouldReload = confirm("Confirm full refresh?");
  e.returnValue = shouldReload;
  return shouldReload;
}

export const usePreventFastRefreshDev = () => {
  useEffect(() => {
    if(process.env.NODE_ENV !== "production") {
      window.addEventListener('beforeunload', confirmPageRefresh);
      return () => window.removeEventListener('beforeunload', confirmPageRefresh);
    }
  }, []);
```

**_app.js**
```
import { usePreventFastRefreshDev } from './some_folder/preventFastRefreshDev';
function MyApp({ Component, pageProps }) {
  usePreventFastRefreshDev();
  <Component {...pageProps} />
}
```

**Comment by peterb0yd** [comment]
It would be amazing to have a config option to turn hot reload off. It's causing unnecessary refreshing in our staging environment.

**Reply by manojgetwealthy** [reply]
I am getting the exact same issue. Not sure why so many unexpected reloads when there is no edit going on with the staging environment.
This is kind of  frustrating.

**Comment by alex-knyazev** [comment]
Hi, Next.js Team! I am triying to disable "fast refresh" too, but can't.

For my solution I have to disable it: I have several https servers ( sites for "n" amount brand) which proxy requests to the same next.js app. 

But issue  is that in such mode "fast refresh" work  for one site only. I tried to go deep in debug and looks like "fast refresh" works only for one server which serves next.js app.

I only found a way to check if request is related to webpack-hmr, and end response, but issue is that page is reloaded in this case which is not what I expect, I want to refresh the page manually.

```js
function handleRequestToHttpsServer( { req, res, nextJsApp, fastRefreshSupported } ) {
    const parsedUrl = parse( req.url, true );

    if ( ! fastRefreshSupported && isFastRefreshRequest( parsedUrl ) ) {
        console.log( req.headers.host );
        endFastRefreshRequestWithPageReloading( res );
    }
    else {
        const brand = getBrandByRequest( req );

        setBrandNameEnvVariable( brand.brandName );

        const handleRequestByNextJsDevServer = nextJsApp.getRequestHandler();

        handleRequestByNextJsDevServer( req, res, parsedUrl );
    }
}

function isFastRefreshRequest( parsedUrl ) {
    const pathName = parsedUrl.pathname;

    return pathName.endsWith( '/webpack-hmr' ) || pathName.endsWith( '.hot-update.json' );
}

function endFastRefreshRequestWithPageReloading( res ) {
    /* In such way I can disable fast-refresh, but page will be reloaded. Can I do something to avoid page-reloading ? Maybe some special response for it? */ 
    res.writeHead( 304);  
    res.end();
}

```

**Reply by meglio** [reply]
Next.js is too busy adding new features, so the core usability issues remain in idle state. I moved to Remix.

**Reply by ScyDev** [reply]
I'll give Remix a try then :)

**Reply by meglio** [reply]
Follow up: Remix was just acquired by Shopify, so expect the quality to drop down and similar stale issues to take place from now on :(

**Comment by lcnlvrz-revelacion** [comment]
This is really frustating when you have npm packages linked locally (npm link ..). Every single change that you made to your external modules causes a full refresh.

Most of the time you have to re-do n interactions to back to last state and it's a waste of time.

**Comment by luisgrases** [comment]
there is a hack you could use in the meantime. If you install https://github.com/preactjs/signals and use the useSignal hook, Fast Refresh stops working. It is currently a bug on their side but could be used as a fix here until a proper solution is found.

**Comment by codewithelvin** [comment]
@nextjsTeam Are you fall in deep sleep? The same problem with latest version of Nextjs. Waiting for updates.

**Comment by hockeydave** [comment]
Caramba... Is there no way to tell nextJS to ignore certain file extensions for a fast refresh?  I am keeping database in a json file.  So ideally I'd want to use something like node ./server.js  --ignore *.json to tell next to not reload on json files.

**Comment by ajguyot** [comment]
For me, Next's HMR worked extremely well until sometime around version 12, when it started intermittently triggering a full reload when I click on links to navigate elsewhere on the website. Initially this only happened occasionally, but the behavior has grown worse and worse with each new version of NextJS. At this point, clicking links while in dev mode will trigger a full refresh the first time, every time. Clicking the link again after the refresh will work, but I have to wait for the slow full-refresh process to complete first, which is terrible. My guess is that it's related to them only trying to optimize for usage of `Next/Link` rather than native `<a>` tag links, but I don't know for sure.

I'm running my web server behind a proxy layer, even locally, so I've actually been able to work around Next's refusal to disable HMR by cutting off its web requests at the proxy. If you disallow the WebSocket URL path `/_next/webpack-hmr` from reaching the NextJS webserver, HMR will immediately stop working in your browser clients, which fixes the constant full-refreshing problem. I use this technique all the time, and then just reenable that URL path when I am doing actual design work and want HMR to function.

The only downside to this approach (other than it requiring a proxy layer, which I presume many devs are not running) is that NextJS kind of freaks out and pollutes your console logs with `WebSocket connection to '/_next/webpack-hmr' failed` messages. They apparently just retry the connection forever, so after about an hour I'll have upwards of 1500 of these errors logged. Thankfully most browsers should automatically group them together, making them pretty easy to ignore.

Maybe someday Next will just allow us to disable HMR, or better yet, will fix whatever problem broke it a couple years ago. In the meantime, hopefully a few of you can save yourselves with this workaround.

**Comment by SilenNaihin** [comment]
+1 would be great to have a way to config this in next.config.json

**Comment by maxymapp** [comment]
_Disclaimer: My solution is not comprehensive and may not work for all use cases..._
FastRefresh will happen if... [The file you're editing might have other exports in addition to a React component.
](https://nextjs.org/docs/architecture/fast-refresh#limitations)
**Workaround**: **If you split your rendered content into custom components, you can edit those component's files without triggering FastRefresh.**
I have a top level server-side rendered page  with `export ... getServerSideProps`.
I extracted some of that rendered content into a separate file `DummyComponent`, and now I can edit `DummyComponent` without FastRefresh, because it has no `export ... getServerSideProps` in it.

**Reply by ScyDev** [reply]
This makes sense. But many people (me included) have the problem, that NextJS does a fast refresh, even if no files are being edited/changed on disk.

**Comment by kiriculum** [comment]
Two years open discussion of a trivial feature to become available? Looks like I picked a wrong week to start a new js framework.
I hope it goes better than chromium issue 5569

**Reply by vanduc1102** [reply]
i have an issue with my project, monolith , nextjs app = `prisma + graphql  +  frontend` , wish there is a way to turn off hot reload

**Comment by mehzee92** [comment]
Sir, next js good but very poor in showing trace of error, file, line
number etc

**Comment by esponges** [comment]
I thought that there was something wrong with my code that was causing the full reload... apparently not.

**Comment by ansu16281** [comment]
will you add this feature? I also need it

**Comment by herawais** [comment]
still waiting...

**Comment by mrgreen33gamer** [comment]
still waiting...

**Comment by niraj-khatiwada** [comment]
I though I was the only one.

**Comment by mamift** [comment]
I couldn't see it referenced so far, but I found another github issue from Dec 2021 where one of the maintainers has indicated they will NOT allow turning off fast refresh/hot reload:
https://github.com/vercel/next.js/issues/13268#issuecomment-992463977
> We are not planning to allow turning off Fast Refresh as it's a critical part of Next.js that we strongly encourage using. 

Yeah OK, it's helpful when it works, but when it doesn't? fast refresh has a bunch of limitations, that they themselves [admit in their docs](https://nextjs.org/docs/architecture/fast-refresh#limitations), but then rule out providing an option to toggle it or an escape hatch when it doesn't work properly or becomes a blocker/encumbrance to development? 

And how is it critical? It's obviously off when deployed to production so how can it be '**critical**' to have?? For crying out loud: **React included in nextjs** _is critical_ - fast refresh is a dev aid that's nice to have but absolutely not **critical** in any common sense definition of the word.

Very disappointing that the maintainers have this arrogant attitude.

**Comment by DucDuong97** [comment]
still waiting...

**Comment by Aniumbott** [comment]
still waiting...

**Comment by DJanocha** [comment]
still waiting....
I can't put in words how frustrating this behaviour is...

EDIT:
looks like non-relevant. actually my settings triggered autosave. details in the next comment.

**Comment by DJanocha** [comment]
Maybe I'm not alone, but for me it happened that In the 
`Users/{username}/Library/Application Support/Code/User/settings.json` (on mac os) I had this setting:
```
  "files.autoSave": "afterDelay"
```
which caused saving in the middle of declaring variables therefore I was seeing a lot of errors in my browser.
When I removed that line and saved the file the problem disappeared.

I'm not sure how this setting got there in the first place, but I saw that just today/yesterday.

For non-mac users:
Maybe just try to find `"files.autoSave": "afterDelay"` in the system (possibly in some `settings.json` file). Copilot told me that default locations for that file would be:
- for windows: C:\Users\{username}\AppData\Roaming\Code\User\settings.json
- for linux: ~/.config/Code/User/settings.json

Copilot conversation for reference:

![image](https://github.com/vercel/next.js/assets/50969285/0a7d9cda-5eee-409d-8c83-53f5e9196533)

Hope that helps,
happy coding 😊

**Comment by gevgeny** [comment]
I solved it by patching next dev server code :( Dirty and fragile solution, but since we don't often reinstall the next module it is still better than nothing: 

1. Go to `node_modules/next/dist/client/components/react-dev-overlay/pages/hot-reloader-client.js`
2. In the `connect ` function, put the initialization under some condition, e.g `localstorage` flag:
```js
...
if (localStorage.getItem('HMR_DISABLED')) {
    console.warn('HMR disabled');
} else {
    // old-code starts
    MODE = mode;
    (0, _client.register)();
    (0, _websocket.addMessageListener)((payload)=>{
        if (!("action" in payload)) {
            return;
        }
        try {
            processMessage(payload);
        } catch (err) {
            var _err_stack;
            console.warn("[HMR] Invalid message: " + payload + "\n" + ((_err_stack = err == null ? void 0 : err.stack) != null ? _err_stack : ""));
        }
    });
    // old-code ends
}
...
```
It works because this code is run on the client.
3. Remove the `.next` folder
4. rerun `next dev` 
5. `localStorage.setItem('HMR_DISABLED', true)`

I did not find a way to pass env variable instead, something like `HMR_DISABLED=true next dev` would be much better. If somebody knows the way, please share.

**Reply by ScyDev** [reply]
Thank you!

**Reply by alexpavlov** [reply]
@gevgeny  Thank you for the workaround. I implemented it by simply commenting out the section to conditionally bypass, and the results were immediate - no more _unexpected_ page reloads, while hot reload still works perfectly.

It’s clear that the current implementation of HMR does not interact well with redirects and rewrites initiated by middleware. The extra requests introduced by this interaction often override navigation and undo the intended redirects/rewrites. This has led to countless hours spent diagnosing _nonexistent_ issues and trying to make redirects/rewrites HMR-friendly.

I believe it’s worth highlighting this for the Next.js team. Obviously they have vision for innovative features, but the rigidity of certain design choices is extremely counterproductive. Making features like this optional or configurable would allow teams to better align the framework with their specific use cases without resorting to workarounds.

Thanks again for sharing the recipe. Спасибо!

**Comment by karldemetria** [comment]
Hello everyone, I encountered the same problem, and this is the solution that worked for me.

I isolated the issue by progressively deleting imports in my layout.tsx file until I reached the globals.css file. When I commented out the lines that included `@tailwind base; @tailwind components; @tailwind utilities;` the Fast Refresh issue suddenly disappeared. However, these Tailwind directives are very important to my app because the layout would be broken without them, so I needed to find a better solution.

I then examined my tailwind.config.js file. I isolated the issue by removing configurations one by one until I discovered that the problem originated from the content configuration line that included `content: ['./src/**/*.{js,ts,jsx,tsx}']` Removing this line made the Fast Refresh issue disappear. Since my directory structure is src/app, I adjusted the content path by adding 'app' before 'src/', resulting in `content: ['./src/app/**/*.{js,ts,jsx,tsx}'],`

This change resolved the Fast Refresh issue without affecting my layout.

I hope this solution works for you all as well.

**Reply by JustCreateDeveloper** [reply]
You're a lifesaver

**Comment by perescope** [comment]
I found that easiest way to turn off HMR is to block `webpack-hmr` url in DevTools. Hope this will help you
<img width="407" alt="image" src="https://github.com/user-attachments/assets/5e49b00b-e6b2-4474-b44e-7ad0ca0cea25">

**Reply by ScyDev** [reply]
Hot damn! This is the way. Someone give this man a medal :)

**Reply by alfi-dim** [reply]
Sadly this not work for me :(
![image](https://github.com/user-attachments/assets/335f41a8-4937-4e4c-848e-b2e2d87ad0be)

**Reply by Frikster** [reply]
Can anyone comment on why this might not work? Are @alfi-dim and I doing something wrong? Can anyone confirm whether this method still works for them? I see the exact same thing as @alfi-dim and it is very frustrating because it shows as blocked but the requests keep occurring.

**Comment by codeaid** [comment]
My HMR suddenly started going crazy without me doing anything to any of the dependencies. It was all working fine for weeks and even this morning and then one moment I get this:

https://github.com/user-attachments/assets/de25b9a7-dc7b-4287-b648-d17be8d33cfd

I've no idea how to fix it. Upgraded and downgraded Next.js. Deleted `node_modules` and reinstalled everything and even blocked the request using the suggested method above. Nothing helps and it just keep going.

**Reply by Frikster** [reply]
I find it so frustrating that deliberately blocking it in chrome-devtools does not in fact block it.

**Comment by goksel-p** [comment]
I turn off the javascript in the tab:

`cmd+shift+p in Chrome console > disable`

![Screenshot by Dropbox Capture](https://github.com/user-attachments/assets/123a1f5b-d426-4a32-881b-f88f4ee5d025)

**Reply by Dan503** [reply]
A decent makeshift solution. Essentially turn it off, save the file, then turn it back on again. The page will not hot-reload and you will still have access to the features that require JS.

**Reply by jay-scribe** [reply]
Absolutely ridiculous we need to disable javascript just to disable hot reload

**Comment by spiridonov-oa** [comment]
Has anyone solved this issue?

**Comment by tybro0103** [comment]
Took time to learn and try out Next.js today. First thing I wanted to do in my hello world app was disable HMR. Being straight-up SOL there is not a great first impression.

**Comment by victormanuelfrancodev** [comment]
I have a bit of a strange problem, and I'm exhausted from trying to fix it. It seems like it's related to this friend [Fast Refresh] rebuilding.
How can I disable this?!! 

I've tried everything, forcing my page to reload every time I enter. Sometimes it works, other times it doesn't, etc. And it's not working. I don't know what to do, I don't know if Next.js is trash or if it's me."

Would you like to make this message more formal or add any specific details for a context where you might share it?

**Comment by gitivon** [comment]
I've noticed that every Server Action call triggers Fast Refresh, even when there are no code changes. This behavior leads to React rendering warnings in the console, specifically:

`Warning: Cannot update a component (`Router`) while rendering a different component`

Expected behavior:
- Server Actions should execute without triggering Fast Refresh when there are no code changes
- The component rendering cycle should remain stable during Server Action calls

Current behavior:
- Every Server Action call triggers Fast Refresh
- This causes unnecessary re-renders and React rendering warnings
- The application's rendering flow is disrupted even when no code has changed

Why does Next.js trigger Fast Refresh on Server Action calls when the code hasn't been modified? Is this intended behavior or a potential issue that needs to be addressed?

**Reply by daveycodez** [reply]
I think this is new to NextJS 15 I just started noticing it :/ Are you using the Pages Router or App Router? I'm on Pages Router and any call to my /api is triggering a Hot Reload. Maybe let's make a separate issue for this so we can get it resolved? I'm also noticing like +100ms latency on my API calls on dev as well

**Reply by daveycodez** [reply]
Remove --turbopack flag in package.json for dev script

**Reply by boiseiben** [reply]
I noticed this too in Next 15.2.4 and 15.2.5 (app router), even when the server action is just an empty function.  Updating to Next 15.3.0 or 15.3.3 seems to have fixed it for me, even leaving the `--turbopack` flag in place.

**Comment by daveycodez** [comment]
Hey just so you guys know, I fixed API requests triggering HMR by removing the turbo flag in npm run dev

**Comment by dahool** [comment]
is there any real solution to this? or still being ignored.

I'm constanly chassing ghosts, I'm never sure if something is working incorrectly due to my fault or due to this useless thing, like for example, all these "fake?" calls to my server side function

` POST /700723065524584599/events 200 in 1014ms
 POST /700723065524584599/events 200 in 1026ms
 POST /700723065524584599/events 200 in 1023ms
 POST /700723065524584599/events 200 in 1019ms
 POST /700723065524584599/events 200 in 1010ms
 POST /700723065524584599/events 200 in 1031ms
 POST /700723065524584599/events 200 in 1015ms
 POST /700723065524584599/events 200 in 1021ms
 POST /700723065524584599/events 200 in 1018ms
 POST /700723065524584599/events 200 in 1018ms
 POST /700723065524584599/events 200 in 1007ms
 POST /700723065524584599/events 200 in 1016ms
 POST /700723065524584599/events 200 in 1020ms
 POST /700723065524584599/events 200 in 1018ms
 POST /700723065524584599/events 200 in 1019ms`

 Yeah, I removed --turbopack, still the same. I had to build and run in production mode to actually determine my code was ok.

**Reply by tybro0103** [reply]
> I'm never sure if something is working incorrectly due to my fault or due to this

It's honestly a really good point that shouldn't be ignored. It doesn't matter how good HMR/fast refresh gets, it's not the same as the code running normally.

It not being optional is nuts.

**Comment by archneon** [comment]
There must be an option to disble automatic refresh when you save! This is crazy. I often make some minimal changes to CSS styles for example and i want to manually switch to the browser, to see the old structure of the page right before i hit refresh, to see if it moved 4px to the right for example. It's almost impossible to see the difference if HMR is auto refreshing. I am not sure who thought this is a good idea. It really depends in which phase of development you are. When i am working with fine tunning CSS ,it's so annoying that it auto refreshes!

**Comment by daveycodez** [comment]
@leerob is there any response from Next team for this issue

**Comment by alex-knyazev** [comment]
heavy plus. I need option to disable fast refresh too. My project is very big, if I just click ctrl+s in my code - all tabs start to fetch new changes, it is heavy long operation, I don't want it, I want to refresh page manually!

**Comment by NexWasTaken** [comment]
Damn 4 years and no response?

**Comment by PiotrBryla** [comment]
It's an issue for my team, too. It would be good to get this `Fast Refresh` configurable.

**Comment by GrahamQuan** [comment]
any update?

**Comment by alcxander** [comment]
getting absolutely frustrated with the code refreshing while developing. why is the default behaviour that as soon as I start typing the site breaks, I don't need big flashy errors in my face while typing literally anything on my IDE. just now started a simple if and the whole site breaks until ive finished typing this, why can't I delay the hot refresh to even wait X seconds after ive at least stopped typing? such an annoying poor poor feature totally not worth the hot reload when it makes me want to regress back to manual load anyway

**Reply by boiseiben** [reply]
It should only change when you save the file. Do you have autosave turned on?

**Comment by arifikhsan** [comment]
still waiting for update 🫠

**Comment by Bespalov-D-A** [comment]
Thanks to the Next.js team — you’re making the world a simpler place.
You’re doing an amazing job, but I was deeply disappointed to find out that there’s no way to disable hot reload in Next.js.

Right now I’m working on a complex application that launches around 10 services — that’s about 18 Docker containers running at once.
Three of them use Next.js on the client side, and after every single file save they all try to reload. My Mac just can’t handle it, even though all the services are well-optimized.

Nowadays, building apps with many services is a pretty common practice. But because we can’t disable hot reload, we’re forced to close and reopen browser tabs just to see updates. That’s just painful.

Why can’t something as simple as an on/off switch be implemented?

**Comment by AstroGD** [comment]
Yes please!
I hate this so much. Turbopack Fast refresh is the most bugged piece of code I had to use ever.
From working fine it has evolved to triggering infinite fast refreshes as soon as I trigger a redirect and now it triggers infinite fast refreshes when I open the Dev tools. HOW?! WHY?!

Please let me turn it off for my sanity's sake, the old builder takes 30+ seconds for each page and I hate being forced to revert to it

**Comment by bhuphathi** [comment]
use solidjs and solidstart no more pain, no reload, no rerender

**Comment by Sbrjt** [comment]
<img width="1088" height="60" alt="image" src="https://github.com/user-attachments/assets/ced65587-4d37-4e11-a1fd-117abfa032c3" />


I can't stand these logs and prefer to hide them in the dev tools filter:

<img width="1099" height="74" alt="image" src="https://github.com/user-attachments/assets/5ba8e0a3-6091-41e4-b3db-fe3225e7c0d3" />

**Comment by LashaMetreveli** [comment]
how is this still an issue....

**Comment by sebastiankade** [comment]
This is absurd

**Comment by emrecpp** [comment]
Not solved, since 2021...

**Comment by kaleb-dev** [comment]
+1

**Comment by mryechkin** [comment]
@timneutkens can you please shed some light on if this is ever going to be an option in Next.js? Ever since the upgrade to version 16 I've been having more and more issues with fast refresh. I don't want to get into why it's not working for me - I just want to have an option to turn it off, and judging by the comments here - so would many others. It would be great to hear from the team if this is ever going to be addressed? Thank you

**Comment by boiseiben** [comment]
Try doing a prod build instead of running the dev server.

**Reply by emrecpp** [reply]
Wow, what a creative idea! 🙄👏

**Comment by zachharris-ef** [comment]
I was able to prevent fast-refresh from being triggered when editing unit tests with this in my `next.config.js`:

```ts
const nextConfig = {
  webpack: config => {
    config.watchOptions: {
      ignored: ["**/__tests__/**"]
    };
    return config;
  }
}
```

I just tested replacing `**/__tests__/**` with `**/*.ts` and `**/*.tsx` and I was able to edit any file without triggering a hot refresh.

**Comment by henrych4** [comment]
It may be off-topic but if your page automatically reloads and there are errors like "WebSocket connection to 'ws://localhost:3000/_next/webpack-hmr' failed" on console, you can check the site permission and update "Apps on device" to "Allowed". This can fix page auto reloading from multiple websock connection failures
<img width="392" height="244" alt="Screenshot 2026-04-17 at 17 47 39" src="https://github.com/user-attachments/assets/bec1c857-b2ce-4e7d-9821-163a9b479fa0" />
